### PR TITLE
[Transforms] Capture `this` in computed property names in arrow functions

### LINKED ISSUE #7860: [Transforms] Missing `this` capture for computed property names inside arrow functions
**Expected**:

```js
   C.prototype.bar = function () {
        var _this = this;
        (function () {
            var obj = (_a = {},
                _a[_this.bar()] = function () { },
                _a);
            var _a;
        });
        return 0;
    };
```

**Actual**:

```js
 C.prototype.bar = function () {
        (function () {
            var obj = (_a = {},
                _a[this.bar()] = function () { } // needs capture
            ,
                _a);
            var _a;
        });
        return 0;
    };
    return C;
```

Failed tests:
* tests/cases/conformance/es6/computedProperties/computedPropertyNames29_ES5.ts 
* tests/cases/conformance/es6/computedProperties/computedPropertyNames31_ES5.ts

### PR DESCRIPTION
**Pull request by sandersn**
Fixes #7860 

Fixes baselines:
1. computedPropertyNames29_ES5
2. computedPropertyNames31_ES5

### CODE REVIEW

File: src/compiler/binder.ts [hunk 1]
```diff
@@ -1945,6 +1951,9 @@ namespace ts {
                     // is an ES6 node.
                     transformFlags = TransformFlags.AssertES6;
                 }
+                if (subtreeFlags & TransformFlags.ContainsLexicalThisInComputedPropertyName) {
```
Comment (yuit) [inline src/compiler/binder.ts]: Could you add comment about the discussion you have with @rbuckton why you convert this to "ContainsLexicalThis"

File: src/compiler/types.ts [hunk 2]
```diff
@@ -2853,10 +2854,10 @@ namespace ts {
         FunctionExcludes = ContainsDecorators | ContainsDefaultValueAssignments | ContainsCapturedLexicalThis | ContainsLexicalThis | ContainsParameterPropertyAssignments | ContainsBlockScopedBinding,
         ConstructorExcludes = ContainsDefaultValueAssignments | ContainsLexicalThis | ContainsCapturedLexicalThis | ContainsBlockScopedBinding,
         MethodOrAccessorExcludes = ContainsDefaultValueAssignments | ContainsLexicalThis | ContainsCapturedLexicalThis | ContainsBlockScopedBinding,
-        ClassExcludes = ContainsDecorators | ContainsPropertyInitializer | ContainsLexicalThis | ContainsCapturedLexicalThis | ContainsComputedPropertyName | ContainsParameterPropertyAssignments,
+        ClassExcludes = ContainsDecorators | ContainsPropertyInitializer | ContainsLexicalThis | ContainsCapturedLexicalThis | ContainsComputedPropertyName | ContainsParameterPropertyAssignments | ContainsLexicalThisInComputedPropertyName,
```
Comment (yuit) [inline src/compiler/types.ts]: Why you need to exclude them? I thought by checking for it in the subTreeFlag then convert it to ContainsLexicalThis, we actually didn't propagate "ContainsLexicalThisInComputedPropertyName"

Comment (sandersn) [inline src/compiler/types.ts]: Nope, the conversion just *adds* to transformFlags -- this list is the part that adds to the list of excludes.

Comment (rbuckton) [inline src/compiler/types.ts]: @yuit, @sandersn  is correct. The `*Excludes` flags are what prevents the propagation. All flags propagate up the tree by default.

**Comment by rbuckton** [comment]
👍 (after adding comments per @yuit above)

File: src/compiler/types.ts (hunk 2 above)
Comment (yuit) [inline src/compiler/types.ts]: gotcha. Thanks

**Comment by yuit** [comment]
👍 after add comment.

**Comment by sandersn** [comment]
OK, it's done, but it is *surprisingly* hard to explain in writing. :)

**Pull request merged**